crates/libs/windows/src/core/event.rs

```diff
@@ -0,0 +1,254 @@
+}
+
+impl<T: Interface + Clone> Default for Array<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: Interface + Clone> Array<T> {
+    /// Creates a new, empty `Array<T>` with no capacity.
+    fn new() -> Self {
+        Self { buffer: std::ptr::null_mut(), len: 0, _phantom: std::marker::PhantomData }
+    }
+    /// Creates a new, empty `Array<T>` with the specified capacity.
+    fn with_capacity(capacity: usize) -> Result<Self> {
+        Ok(Self { buffer: Buffer::new(capacity * std::mem::size_of::<Delegate<T>>())?, len: 0, _phantom: std::marker::PhantomData })
+    }
+    /// Swaps the contents of two `Array<T>` objects.
+    fn swap(&mut self, mut other: Self) -> Self {
+        unsafe { std::ptr::swap(&mut self.buffer, &mut other.buffer) };
+        std::mem::swap(&mut self.len, &mut other.len);
+        other
+    }
+    /// Returns `true` if the array contains no delegates.
+    fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+    /// Returns the number of delegates in the array.
+    fn len(&self) -> usize {
+        self.len
+    }
+    /// Appends a delegate to the back of the array.
+    fn push(&mut self, delegate: Delegate<T>) {
+        unsafe {
+            std::ptr::write((*self.buffer).as_mut_ptr::<Delegate<T>>().add(self.len) as _, delegate);
+            self.len += 1;
+        }
+    }
+    /// Returns a slice containing of all delegates.
+    fn as_slice(&self) -> &[Delegate<T>] {
+        if self.is_empty() {
+            &[]
+        } else {
+            unsafe { std::slice::from_raw_parts((*self.buffer).as_ptr::<Delegate<T>>() as _, self.len) }
+        }
+    }
+    /// Returns a mutable slice of all delegates.
+    fn as_mut_slice(&mut self) -> &mut [Delegate<T>] {
+        if self.is_empty() {
+            &mut []
+        } else {
+            unsafe { std::slice::from_raw_parts_mut((*self.buffer).as_mut_ptr::<Delegate<T>>() as _, self.len) }
+        }
+    }
+}
+
+impl<T: Interface + Clone> Clone for Array<T> {
+    fn clone(&self) -> Self {
+        if !self.is_empty() {
+            unsafe { (*self.buffer).0.add_ref() };
+        }
+        Self { buffer: self.buffer, len: self.len, _phantom: std::marker::PhantomData }
+    }
+}
+
+impl<T: Interface + Clone> Drop for Array<T> {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.is_empty() && (*self.buffer).0.release() == 0 {
+                std::ptr::drop_in_place(self.as_mut_slice());
+                heap_free(self.buffer as _)
+            }
+        }
+    }
+}
+
+/// A reference-counted buffer.
+#[repr(C)]
+struct Buffer(RefCount);
+
+impl Buffer {
+    /// Creates a new `Buffer` with the specified size in bytes.
+    fn new(size: usize) -> Result<*mut Buffer> {
+        if size == 0 {
+            Ok(std::ptr::null_mut())
+        } else {
+            let alloc_size = std::mem::size_of::<Buffer>() + size;
+            let header = heap_alloc(alloc_size)? as *mut Buffer;
```

Comment:
The other code makes assumptions that this pointer is properly aligned. Since `Buffer` is an `AtomicI32` can we assume that the pointer will always be properly aligned?

Reply:
Yes, `HeapAlloc` returns memory is aligned to 8 bytes on x86 and 16 bytes on x64.